src/sync.rs

```diff
@@ -0,0 +1,302 @@
+    fn mutex_enqueue(&mut self, id: MutexId, thread: ThreadId) {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.push_back(thread);
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_dequeue(&mut self, id: MutexId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.pop_front()
+    }
+
+    #[inline]
+    /// Create state for a new read write lock.
+    fn rwlock_create(&mut self) -> RwLockId {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks.push(Default::default())
+    }
+
+    #[inline]
+    /// Check if locked.
+    fn rwlock_is_locked(&mut self, id: RwLockId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.is_some()
+            || !this.machine.threads.sync.rwlocks[id].readers.is_empty()
+    }
+
+    #[inline]
+    /// Check if write locked.
+    fn rwlock_is_write_locked(&mut self, id: RwLockId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.is_some()
+    }
+
+    /// Add a reader that collectively with other readers owns the lock.
+    fn rwlock_reader_add(&mut self, id: RwLockId, reader: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(!this.rwlock_is_write_locked(id), "the lock is write locked");
+        let count = this.machine.threads.sync.rwlocks[id].readers.entry(reader).or_insert(0);
+        *count += 1;
+    }
+
+    /// Try removing the reader. Returns `true` if succeeded.
+    fn rwlock_reader_remove(&mut self, id: RwLockId, reader: ThreadId) -> bool {
+        let this = self.eval_context_mut();
+        match this.machine.threads.sync.rwlocks[id].readers.entry(reader) {
+            Entry::Occupied(mut entry) => {
+                let count = entry.get_mut();
+                *count -= 1;
+                if *count == 0 {
+                    entry.remove();
+                }
+                true
+            }
+            Entry::Vacant(_) => false,
+        }
+    }
+
+    #[inline]
+    /// Put the reader in the queue waiting for the lock.
+    fn rwlock_enqueue_reader(&mut self, id: RwLockId, reader: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(this.rwlock_is_write_locked(id), "queueing on not write locked lock");
+        this.machine.threads.sync.rwlocks[id].reader_queue.push_back(reader);
+    }
+
+    #[inline]
+    /// Take the reader out the queue waiting for the lock.
+    fn rwlock_dequeue_reader(&mut self, id: RwLockId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].reader_queue.pop_front()
+    }
+
+    #[inline]
+    /// Lock by setting the writer that owns the lock.
+    fn rwlock_writer_set(&mut self, id: RwLockId, writer: ThreadId) {
```

Comment:
Again terminology here is a bit confusing, this too should IMO be "lock" or "acquire (lock)".